sim/midas/src/main/scala/midas/models/sram/BankedAsyncMemModel.scala

```diff
@@ -0,0 +1,138 @@
+
+  // Structures to organize read/write ops
+  case class Read(cmd: DecoupledIO[ReadCmd], resp: DecoupledIO[UInt], done: Bool) {
+    def fireable = cmd.valid && resp.ready && !done
+    def skippable = fireable && !cmd.bits.en && hostParams.enableReadSkips.B
+  }
+
+  case class Write(cmd: DecoupledIO[WriteCmd], done: Bool) {
+    def fireable = cmd.valid && !done
+    def skippable = fireable && !(cmd.bits.en && cmd.bits.mask) && hostParams.enableWriteSkips.B
+  }
+
+  // Use up to two underlying banks of TDP BRAMs
+  val nPhysReads = hostParams.maxPhysReads.min(nReads)
+  val physReadCmds = Wire(Vec(nPhysReads, new ReadCmd(depth, dataWidth)))
+  val physReadResps = Wire(Vec(nPhysReads, UInt(dataWidth.W)))
+
+  // All writes (up to two) get broadcast to all banks
+  val nPhysWrites = hostParams.maxPhysWrites.min(nWrites)
+  val physWriteCmds = Wire(Vec(nPhysWrites, new WriteCmd(depth, dataWidth)))
+
+  // Put each target read's metadata into a product for convenience
+  val reads = (channels.read_cmds zip channels.read_resps).map {
+    case (rc, rr) =>
+      val rDone = RegInit(false.B)
+      rDone.suggestName("rDone")
+      Read(rc, rr, rDone)
+  }
+
+  // Divide target reads as evenly as possible into nPhysReads banks
+  val rBankMaxSize = (channels.read_cmds.length + nPhysReads - 1) / nPhysReads
+
+  // Pick the winning read for each read bank
+  //   - Hook it up to the underlying memory's corresponding port
+  //   - Mark it done
+  (reads.grouped(rBankMaxSize).toSeq zip (physReadCmds zip physReadResps)).foreach {
+    case (rBank, (underlyingRC, underlyingRR)) =>
+      underlyingRC.en := true.B
+      underlyingRC.addr := rBank.head.cmd.bits.addr
+      val activeOH = chisel3.util.PriorityEncoderOH(rBank.map(_.fireable))
+      (rBank zip activeOH).foreach {
+        case (r, active) =>
+          val active_piped = RegNext(active)
+          r.cmd.ready := active
+          r.resp.valid := active_piped
+          r.resp.bits := underlyingRR
+          when (active) {
+            underlyingRC.addr := r.cmd.bits.addr
+            r.done := true.B
+          }
+       }
+  }
+
+  // Writes can only begin once all reads are done
+  val allReadsDone = reads.map(_.done).reduce(_ && _)
+
+  // Put each target write's metadata into a product for convenience
+  val writes = channels.write_cmds.map {
+    case wc =>
+      val wDone = RegInit(false.B)
+      wDone.suggestName("wDone")
+      Write(wc, wDone)
+  }
+
+  // Divide target writes as evenly as possible into nPhysWrites banks
+  val wBankMaxSize = (channels.write_cmds.length + nPhysWrites - 1) / nPhysWrites
+
+  // Pick the winning write for each write bank
+  //   - Hook it up to the underlying memory's corresponding port
+  //   - Mark it done
+  (writes.grouped(wBankMaxSize).toSeq zip physWriteCmds).foreach {
+    case (wBank, underlyingWC) =>
+      underlyingWC.en := false.B
+      underlyingWC.mask := true.B
```

Comment:
Is this not overridden somewhere? If it's unsupported you should add some assertions.

Reply:
What thing being unsupported are you referring to?

Reply:
The mask

Reply:
The target mask is supported.

Reply:
Keep in mind these are ground-type memories. The scheduler picks a write only when it is active (mask and en both true) and skips it otherwise, so the value of the target mask is implemented correctly; however, the mask of the internal host memory is always set to true. This makes both false-mask and false-enable ports "skippable," which is some trivial performance benefit.

Reply:
Duh, ok. It might be worth just add a small comment noting that.

Reply:
I added a comment.